fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java

```diff
@@ -800,57 +833,121 @@ public CompletableFuture<RecordCursorContinuation> rebalancePartitions(RecordCur
-                                return Pair.of(
-                                        movedCount,
-                                        Math.max(partitionInfo.getCount() - movedCount - indexPartitionHighWatermark, 0));
-                            });
+                    if (previous != null && previous.getCount() + currentPartitionCount <= indexPartitionHighWatermark) {
+                        // merge current into previous
+                        repartitioningContext.cursor = getOldestNDocuments(partitionInfo, groupingKey, repartitioningContext.countToMove);
+                    } else if (next != null && next.getCount() + currentPartitionCount <= indexPartitionHighWatermark) {
+                        // merge current into next
+                        repartitioningContext.removingOldest = false;
+                        repartitioningContext.cursor = getNewestNDocuments(partitionInfo, groupingKey, repartitioningContext.countToMove);
+                    } else if (previous != null && next != null && (2 * indexPartitionHighWatermark - previous.getCount() - next.getCount()) >= currentPartitionCount ) {
+                        // split current into next and previous
+                        // for this iteration, we do a partial move into previous; next iteration(s) would eventually
+                        // move the rest to next
+                        int previousCapacity = indexPartitionHighWatermark - previous.getCount();
+                        repartitioningContext.countToMove = Math.min(previousCapacity, repartitionDocumentCount);
+                        if (currentPartitionCount > repartitioningContext.countToMove) {
+                            repartitioningContext.countToMove++;
+                            repartitioningContext.newBoundaryRecordPresent = true;
+                        }
+                        repartitioningContext.emptyingPartition = false;
+                        repartitioningContext.cursor = getOldestNDocuments(partitionInfo, groupingKey, repartitioningContext.countToMove);
+                    } else {
+                        // partition is under low-watermark, but no capacity around it to absorb it
+                        continue;
+                    }
+                } else {
+                    // partition is above high watermark
+                    repartitioningContext.countToMove = 1 + Math.min(repartitionDocumentCount, indexPartitionHighWatermark);
+                    // sanity: in case someone passes a high enough repartitionDocumentCount such that removing
+                    // this many docs would flip the partition from being above the high watermark to being below
+                    // the low watermark
+                    if (currentPartitionCount - repartitioningContext.countToMove < indexPartitionLowWatermark) {
+                        repartitioningContext.countToMove = indexPartitionHighWatermark - indexPartitionLowWatermark;
+                    }
```

Comment:
I guess it's probably a bug if this happens, so maybe optimizing it is irrelevant, but it seems to me that you would want to move `(indexPartitionHighWatermark + indexPartitionLowWatermark) / 2`, rather than bringing it all the way to the low watermark.

Reply:
you mean `(indexPartitionHighWatermark - indexPartitionLowWatermark) / 2` ?